### PR TITLE
Query: Perf improvements to query annotations.

### PR DESCRIPTION
**Pull request by anpete**
- Use Result Operators for query annotations. Avoids an ET rewrite during query cache resolution.
- FromSql queries correctly handled in query cache.
- Fixed some issues with null semantics processing. Always do null semantics rewriting during SQL gen.
- Streamlined and standardized some ET visitor code.

### CODE REVIEW

File: src/EntityFramework.Core/Query/ResultOperators/Internal/TrackingExpressionNode.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            : base(parseInfo, null, null)
+        {
+        }
+
+        protected override ResultOperatorBase CreateResultOperator(ClauseGenerationContext clauseGenerationContext)
+            => new TrackingResultOperator(
+                tracking: ParsedExpression.Method.GetGenericMethodDefinition()
+                          == EntityFrameworkQueryableExtensions.AsTrackingMethodInfo)
+            {
+                //QuerySource =
```
Comment (maumar) [inline src/EntityFramework.Core/Query/ResultOperators/Internal/TrackingExpressionNode.cs]: remove

Comment (anpete) [inline src/EntityFramework.Core/Query/ResultOperators/Internal/TrackingExpressionNode.cs]: Will do.

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/CompositePredicateExpressionVisitor.cs [hunk 2]
```diff
@@ -19,72 +19,22 @@ public CompositePredicateExpressionVisitor(bool useRelationalNulls)
 
         public override Expression Visit([NotNull] Expression expression)
         {
-            var currentExpression = expression;
-            var inExpressionOptimized =
-                new EqualityPredicateInExpressionOptimizer().Visit(currentExpression);
+            expression = new EqualityPredicateInExpressionOptimizer().Visit(expression);
 
-            currentExpression = inExpressionOptimized;
+            var predicateNegationExpressionOptimizer = new PredicateNegationExpressionOptimizer();
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/CompositePredicateExpressionVisitor.cs]: new PredicateNegationExpressionOptimizer().Visit(expression) for consistency with above

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsExpandingVisitor.cs [hunk 3]
```diff
@@ -9,114 +9,102 @@ namespace Microsoft.Data.Entity.Query.ExpressionVisitors.Internal
...
                 }
             }
 
-            if (left == expression.Left
-                && right == expression.Right)
+            if (newLeft != binaryExpression.Left
+                || newRight != binaryExpression.Right)
             {
-                return expression;
+                return Expression.MakeBinary(binaryExpression.NodeType, newLeft, newRight, binaryExpression.IsLiftedToNull, binaryExpression.Method);
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsExpandingVisitor.cs]: break down into multiple lines

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs [hunk 4]
```diff
@@ -7,78 +7,66 @@ namespace Microsoft.Data.Entity.Query.ExpressionVisitors.Internal
...
             }
 
-            if (left == expression.Left
-                && right == expression.Right)
-            {
-                return expression;
-            }
-
-            return Expression.MakeBinary(expression.NodeType, left, right, expression.IsLiftedToNull, expression.Method);
+            return newLeft != binaryExpression.Left
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs]: apparently this check is built into the Update method, so we could remove it

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs [hunk 5]
```diff
@@ -7,78 +7,66 @@ namespace Microsoft.Data.Entity.Query.ExpressionVisitors.Internal
 {
     public class RelationalNullsOptimizedExpandingVisitor : RelationalNullsExpressionVisitorBase
     {
-        public virtual bool OptimizedExpansionPossible { get; set; }
+        public virtual bool IsOptimalExpansion { get; private set; } = true;
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs]: not so sure we are doing *optimal* expansion for all cases where this is true

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsExpandingVisitor.cs (hunk 3 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsExpandingVisitor.cs]: Will change. Changed some others to Update but missed that.

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs (hunk 4 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs]: Will change. Changed some others to Update but missed that.

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs (hunk 5 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs]: It's relative here :-)

File: src/EntityFramework.Core/Query/ResultOperators/Internal/IncludeResultOperator.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+using Microsoft.Data.Entity.Internal;
+using Remotion.Linq;
+using Remotion.Linq.Clauses;
+using Remotion.Linq.Clauses.Expressions;
+using Remotion.Linq.Clauses.ResultOperators;
+using Remotion.Linq.Clauses.StreamedData;
+
+namespace Microsoft.Data.Entity.Query.ResultOperators.Internal
+{
+    public class IncludeResultOperator : SequenceTypePreservingResultOperatorBase, IQueryAnnotation
```
Comment (mikary) [inline src/EntityFramework.Core/Query/ResultOperators/Internal/IncludeResultOperator.cs]: Is the functionality of the result operator and the query annotation related? It seems like there should be a separation of concerns unless there is a benefit to this kind of factoring (i.e. updating the expression updates the annotations)

Comment (anpete) [inline src/EntityFramework.Core/Query/ResultOperators/Internal/IncludeResultOperator.cs]: Yes, the result operator is the implementation of the annotation. This simplifies things and also means that we don't incur an ET rewrite during query cache resolution.

**Comment by mikary** [comment]
:+1: LGTM

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs [hunk 7]
```diff
@@ -7,78 +7,63 @@ namespace Microsoft.Data.Entity.Query.ExpressionVisitors.Internal
...
-                OptimizedExpansionPossible = false;
+                IsOptimalExpansion = false;
             }
 
-            if (operand == expression.Operand)
-            {
-                return expression;
-            }
-            return Expression.MakeUnary(expression.NodeType, operand, expression.Type);
+            return operand != expression.Operand
```
Comment (maumar) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs]: this check can also be removed

**Comment by maumar** [comment]
:shipit:  after the last reducnat check before Update is removed

File: src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs (hunk 7 above)
Comment (anpete) [inline src/EntityFramework.Relational/Query/ExpressionVisitors/Internal/RelationalNullsOptimizedExpandingVisitor.cs]: Done

**Pull request merged**